### PR TITLE
MEPTS 1348 | TX RTT Changes for December 2023 Release

### PR DESCRIPTION
**Pull request by TinoMuianga**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java [hunk 1]
```diff
@@ -555,51 +497,589 @@ public CohortDefinition getTreatmentInterruptionOfXDaysBeforeReturningToTreatmen
...
+   *   <li>Drug pick-up date registered on (FILA)
+   *   <li>Drug pick-up date registered on (Recepção Levantou ARV) – Master Card
+   * </ul>
+   *
+   * @return String
+   */
+  public CohortDefinition getPatientsFirstDrugPickup() {
+    SqlCohortDefinition cd = new SqlCohortDefinition();
+    cd.setName("Patient’s earliest drug pick-up");
+    cd.addParameter(new Parameter("startDate", "Start Date", Date.class));
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java]: Start date is never used

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java [hunk 2]
```diff
@@ -555,51 +497,589 @@ public CohortDefinition getTreatmentInterruptionOfXDaysBeforeReturningToTreatmen
...
+            + " WHERE start.first_pickup >= "
+            + artStartPeriod
+            + " AND start.first_pickup < :startDate";
+
+    cd.setQuery(query);
+    return cd;
+  }
+
+  /**
+   * Patients whose first ever drug pick-up date between the following sources falls during the
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java:571]: FR 4 The pickup falls by the end of previous reporting period

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java [hunk 3]
```diff
@@ -555,51 +497,589 @@ public CohortDefinition getTreatmentInterruptionOfXDaysBeforeReturningToTreatmen
...
+    cd.addSearch(
+        "earliestArtStartDateAfterPeriod",
+        EptsReportUtils.map(earliestArtStartDateAfterPeriod, DEFAULT_MAPPING));
+    cd.addSearch(
+        "firstDrugPickUpAfterPeriod",
+        EptsReportUtils.map(firstDrugPickUpAfterPeriod, DEFAULT_MAPPING));
+
+    cd.setCompositionString(
+        "earliestArtStartDateBeforePeriod OR (earliestArtStartDateAfterPeriod AND firstDrugPickUpAfterPeriod)");
+    return cd;
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java]: FR4 - The pickup falls by the end of previous reporting period

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java [hunk 4]
```diff
@@ -555,51 +497,589 @@ public CohortDefinition getTreatmentInterruptionOfXDaysBeforeReturningToTreatmen
...
+    cd.addParameter(new Parameter("startDate", "Start Date", Date.class));
+    cd.addParameter(new Parameter("endDate", "End Date", Date.class));
+    cd.addParameter(new Parameter("location", "Location", Location.class));
+
+    CommonQueries commonQueries = new CommonQueries(new CommonMetadata(), new HivMetadata());
+
+    String query =
+        "       SELECT patient_id "
+            + " FROM ( "
+            + commonQueries.getARTStartDate(true)
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java:538]: You are mapping report end date to Art start cohort, but according to the requirement the date falls by previous period

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java [hunk 5]
```diff
@@ -555,51 +497,589 @@ public CohortDefinition getTreatmentInterruptionOfXDaysBeforeReturningToTreatmen
...
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java:560]: You are mapping report end date to Art start cohort, but according to the requirement the date falls by previous period

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java [hunk 6]
```diff
@@ -555,51 +497,589 @@ public CohortDefinition getTreatmentInterruptionOfXDaysBeforeReturningToTreatmen
...
+    SqlCohortDefinition cd = new SqlCohortDefinition();
+    cd.setName("Patient’s earliest drug pick-up");
+    cd.addParameter(new Parameter("startDate", "Start Date", Date.class));
+    cd.addParameter(new Parameter("endDate", "End Date", Date.class));
+    cd.addParameter(new Parameter("location", "Location", Location.class));
+
+    String query =
+        "       SELECT patient_id "
+            + " FROM ( "
+            + commonQueries.getFirstDrugPickup()
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java]: You are mapping report end date to first drug pickup cohort, but according to the requirement the date falls by previous period

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java [hunk 7]
```diff
@@ -555,51 +497,589 @@ public CohortDefinition getTreatmentInterruptionOfXDaysBeforeReturningToTreatmen
...
+            + "                GROUP  BY p.patient_id) most_recent "
+            + "        GROUP  BY most_recent.patient_id) final "
+            + "GROUP  BY final.patient_id";
+
+    StringSubstitutor sb = new StringSubstitutor(map);
+
+    cd.setQuery(sb.replace(query));
+
+    return cd;
+  }
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java]: Does not meet the requirements

**Review by eliazmuchave — CHANGES_REQUESTED** [review]
Left comments. Take a look

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java (hunk 7 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java]: This i d done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java (hunk 6 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java:596]: Done, mapped for previous reporting date

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java (hunk 5 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java:565]: Done, mapped for previous reporting date

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java [hunk 8]
```diff
@@ -555,51 +500,629 @@ public CohortDefinition getTreatmentInterruptionOfXDaysBeforeReturningToTreatmen
...
+   * Absolute CD4 Count
+   *
+   * @return CohortDefinition
+   */
+  public CohortDefinition getCd4Result(
+      AdvancedDiseaseAndTBCascadeCohortQueries.Cd4CountComparison cd4CountComparison) {
+
+    SqlCohortDefinition cd = new SqlCohortDefinition();
+    cd.setName("Patients with CD4 Result");
+    cd.addParameter(new Parameter("startDate", "Start Date", Date.class));
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java]: Never used

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java [hunk 9]
```diff
@@ -555,51 +501,634 @@ public CohortDefinition getTreatmentInterruptionOfXDaysBeforeReturningToTreatmen
...
+            + "                                  AND o.concept_id = ${23866} "
+            + "                                  AND o.value_datetime IS NOT NULL "
+            + "                                  AND o.value_datetime < :startDate "
+            + "                                  AND e.location_id = :location "
+            + "                           GROUP  BY p.patient_id) most_recent "
+            + "                   GROUP  BY most_recent.patient_id) "
+            + "                  most_recent_of_previous_period "
+            + "               ON most_recent_of_previous_period.patient_id = p.patient_id "
+            + "WHERE p.voided = 0 "
+            + "AND  Timestampdiff(DAY, restarted.pick_up_date, "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxRttCohortQueries.java]: Check param order

**Review by eliazmuchave — CHANGES_REQUESTED** [review]

**Review by eliazmuchave — APPROVED** [review]
Looks good to me

**Pull request merged**